src/v/resource_mgmt/available_memory.h

```diff
@@ -0,0 +1,129 @@
+    available_memory() = default;
+    available_memory(available_memory&) = delete;
+    available_memory& operator=(const available_memory&) = delete;
+
+    /**
+     * @brief Return the current available memory value.
+     *
+     * Available memory is free memory, plus all the memory that can be in
+     * princple reclaimed from subsystems which have registered a reclaim hook.
+     *
+     * Effectively, this is get_reclaimable() +
+     * ss::memory::stats().free_memory()
+     *
+     * @return size_t the estimated amount of available memory in the system
+     */
+    size_t get_available() const;
+
+    /**
+     * This method calculates the current reclaimable by polling all the
+     * registered reclaimers for how much reclaimable memory they hold, and
+     * returning the sum of all of these amounts.
+     *
+     * The returned value is the assumed amount of memory that could be released
+     * back to the allocator if all reclaimers did an exhaustive reclaim.
+     */
+    size_t get_reclaimable() const;
+
+    /**
+     * @brief Register associated metric(s) for available memory.
+     *
+     * This call registers an "available_memory" metric in the default
+     * metric groups, which reports the get_available() call when it is
+     * polled. Without this call, the class still works fine, exposing
+     * its API to callers, but no metrics are created.
+     */
+    void register_metrics();
+
+    /**
+     * @brief Register an available memory reporter.
+     *
+     * A reporter is a component that can reclaim memory, and so needs to
+     * report the amount of reclaimable memory it holds.
+     *
+     * This method returns a deregister_holder, which when destroyed
+     * deregisters this reporter. Typically, this object will be stored as
+     * a memory of the object to which the reporter function points so that
+     * the function does not access a dangling reference after it is destroyed.
+     *
+     * @param name name of the reporter
+     * @param reporter_fn a function which returns the amount of reclaimable
+     * memory held at any moment in time by this reporter
+     * @return an opaque object that deregisters the reporter
+     */
+    template<typename F>
+    [[nodiscard("You need to hold the returned object to maintain "
+                "registration")]] deregister_holder
+    register_reporter(const ss::sstring& name, F reporter_fn) {
+        return inner_register(name, std::move(reporter_fn));
+    }
+
+    /**
+     * @brief Get a reference to the shard-global available_memory instance.
+     *
+     * @return The global available_memory object for this shard.
+     */
+    static available_memory& local() { return _local_instance; }
+
+private:
+    deregister_holder inner_register(const ss::sstring& name, afn&& fn);
+
+    static thread_local available_memory _local_instance;
+
+    intrusive_list<reporter, &reporter::hook> _reporters;
+    std::optional<ss::metrics::metric_groups> _metrics;
```

Comment:
We should use public metrics for something high value and low cardinality like this

see e.g. where storage::probe does it:
``    ss::metrics::metric_groups _public_metrics{
      ssx::metrics::public_metrics_handle};`
```

Reply:
Fair enough. I do find all this boilerplate we have to do to add a metric to 2 places a bit awkward. We should wrap it up into a new `metric_group`-like object which puts it in both places.

Reply:
I have added the metrics to public metrics.